src/sctools/test/test_count.py

```diff
@@ -1,44 +1,578 @@
+        _test_gene_expression_rate)
+
+    with tempfile.TemporaryDirectory() as _test_temp_dir:
+        # generate test data
+        synthetic_data_generator.generate_synthetic_bam_and_counts_matrix(
+            _test_temp_dir,
+            _test_num_duplicates,
+            _test_num_missing_some_tags,
+            _test_num_multiple_gene_alignments,
+            _test_max_gene_hits_per_multiple_gene_alignments,
+            query_sort_order=query_sort_order)
+
+        # test data paths
+        test_bam_path = os.path.join(_test_temp_dir, SyntheticBarcodedBAMGenerator.bam_output_filename)
+        test_count_matrix_path = os.path.join(_test_temp_dir, SyntheticBarcodedBAMGenerator.count_matrix_output_filename)
+        test_row_index_path = os.path.join(_test_temp_dir, SyntheticBarcodedBAMGenerator.row_index_output_filename)
+        test_col_index_path = os.path.join(_test_temp_dir, SyntheticBarcodedBAMGenerator.col_index_output_filename)
 
-_bam_file = _data_dir + '/cell-sorted2.bam'
-_gtf_file = _data_dir + '/chr1.gtf.gz'
+        # create CountMatrix from the synthetic bam
+        count_matrix_from_bam: CountMatrix = CountMatrix.from_sorted_tagged_bam(test_bam_path, _test_annotation_file)
 
+        # load the test counts matrix
+        count_matrix_data_expected = np.load(test_count_matrix_path)
+        row_index_expected = np.load(test_row_index_path)
+        col_index_expected = np.load(test_col_index_path)
 
-# todo this test should be made faster
-def test_count_matrix():
-    count_matrix = CountMatrix.from_bam(_bam_file, _gtf_file)
-    assert count_matrix._matrix.shape == (734, 3132)
+    count_matrix_data_from_bam = count_matrix_from_bam.matrix.todense()
+    row_index_from_bam = count_matrix_from_bam.row_index
+    col_index_from_bam = count_matrix_from_bam.col_index
 
-    # check save + load
-    count_matrix.save(_test_dir + '/test_count_matrix')
-    count_matrix_2 = CountMatrix.load(_test_dir + '/test_count_matrix')
+    # sort expected and from_bam results by their respective row and column indices, since their sort order
+    # is not part of the design specs and is considered arbitrary
+    (sorted_count_matrix_data_from_bam,
+     sorted_row_index_from_bam,
+     sorted_col_index_from_bam) = _get_sorted_count_matrix(
+        count_matrix_data_from_bam, row_index_from_bam, col_index_from_bam)
+    (sorted_count_matrix_data_expected,
+     sorted_row_index_expected,
+     sorted_col_index_expected) = _get_sorted_count_matrix(
+        count_matrix_data_expected, row_index_expected, col_index_expected)
 
-    # check matrices are the same
-    csr1: sp.csr_matrix = count_matrix._matrix
-    csr2: sp.csr_matrix = count_matrix_2._matrix
-    assert np.allclose(csr1.indices, csr2.indices)
-    assert np.allclose(csr1.indptr, csr2.indptr)
-    assert np.allclose(csr1.data, csr2.data)
+    # assert equality
+    assert np.allclose(sorted_count_matrix_data_from_bam, sorted_count_matrix_data_expected)
```

Comment:
Can we add a docstring to this test that's super clear on what we're testing here? My initial concern was that we're generating synthetic bams and count matrices, and using similar code to transition between them, instead of using some kind of external gold standard that we're aiming to hit. 

Done well, these tests are probably way better than some small matrix or bam used to smoke test the functionality, but there's also an opportunity to code `2 + 2 = 5` somewhere into both the count matrix generation and the bam_to_count function. 

[moonshot] Longer term, when we've figured out all the discrepancies, what do you think about trying to run the cell ranger code in our tests to verify we're equivalent?

Reply:
Added an extensive preamble documentation block.